### ISSUE TITLE
[RFE] Enable Azure accelerated networking - RKE2/K3s

### ISSUE THREAD

**Issue by gaktive** [rancher/dashboard#7182]
Internal reference: SURE-3197
Reported in 2.5.9

Request description:
Node templates should have the option to create VMs in Azure public cloud with accelerated networking enabled.

Here's a Microsoft post about this feature: https://azure.microsoft.com/fr-fr/blog/maximize-your-vm-s-performance-with-accelerated-networking-now-generally-available-for-both-windows-and-linux/

Actual behavior:
It's not an option

Expected behavior:
Should be one

Additional notes from @HarrisonWAffel:

> I've taken a look at what it would take to implement this enhancement and it seems the following changes would be needed. Each change is accompanied by a story point estimate. Each point represents one day of development work, and does not include time needed for code review or QA verification.

### Dashboard Changes
- RKE1 UI would need a new check-box in the Azure node template, in the 'Networking' Section. Enabling this checkbox would enable accelerated networking. 
- RKE2 UI would need a new check box in the cluster creation screen, under the 'Advanced Options' for each node pool.
- The default machine config would need to be updated to include the new option for accelerated networking, it should be named 'acceleratedNetworking' and its default value should be 'false'

(There would be required Machine and Terraform changes too.)

> Some VM options do not support accelerated networking, though most do. The Azure API provides explanatory error messages which are bubbled up to the UI, so a user should be able to self diagnose their configuration if they select an incompatible VM. Azure does not provide a static list of VM's supporting accelerated networking, however they do have documentation on how to determine support using their CLI tool https://learn.microsoft.com/en-us/azure/virtual-network/accelerated-networking-overview#supported-vm-instances.

Connection to rancher/dashboard#7177 or in Zube, rancher/dashboard#7177

**Comment by catherineluse** [comment]
When the UI gets the list of Azure VM sizes, it comes in the format of a list of about 600 strings originating from `https://github.com/rancher/rancher/blob/release/v2.7/pkg/api/norman/customization/aks/listers.go#L376`. Since we are adding the accelerated networking and availability zone options on 2.7.1, I wonder if it would be feasible to have the UI pass query options to that same endpoint in order to filter the VMs by the selected availability zone and whether they support accelerated networking or not, or if it would be better for the UI to bypass that endpoint and call the Azure API separately.

If the user has an existing machine pool of a certain size, then selects a region that doesn't support it, they should see an error message that warns them that it isn't supported. On the other hand, if the user is creating a new machine pool and has selected an availability zone, then the VM list should automatically be filtered to only show VMs available in that region. But because hardcoded lists get outdated so quickly, this only works if the UI can do those queries to determine whether a certain combination of options is supported.

I started looking into capturing the API request made by the command for checking the availability of accelerated networking in the Azure docs:

```
 az vm list-skus -l westus --all true -r virtualMachines --query '[].{size:size, name:name, acceleratedNetworkingEnabled: capabilities[?name==`AcceleratedNetworkingEnabled`].value | [0]}' -o table
```

In theory, I believe the UI could make those types of query requests to the Azure API without backend changes due to the fact that we can use Rancher's proxy (https://github.com/rancher/rancher/blob/release/v2.6/pkg/httpproxy/proxy.go) to make requests to third-party domains. So I've started looking into it. I'm not sure if anyone feels strongly that these kinds of queries should be handled on the back end.

**Comment by catherineluse** [comment]
@HarrisonWAffel After looking into it a bit more, I found that the UI cannot get the VM sizes from the Azure API because in order to access that endpoint, the UI needs a bearer token which can only be created using the tenantId, clientId and clientSecret from the Azure cloud credentials. I assumed something like this would be public knowledge but it requires authentication.

When the UI gets a cloud credential from Norman at the endpoint `v3/cloudcredentials`, the tenantId and clientSecret are not included in the response. I assume that's for security reasons, so in order to support the filtered VM sizes, the backend would need to provide the VM sizes with the extra data attached (location, availability zone, accelerated networking compatibility, and compatibility with managed disks).

**Comment by HarrisonWAffel** [comment]
@catherineluse We can expand the [existing Norman API](https://github.com/rancher/rancher/blob/625b6107bb1dd55fdcc8d979c761a31069e3fb39/pkg/api/norman/customization/aks/handler.go#L134) with a new path that will include this information by listing the capabilities of each VM SKU (VM type / size), using the same Azure API endpoint you've included in your above comment. This API is currently used by the UI when creating RKE2/K3s clusters to list the available VM sizes. With a new path we can give the UI the needed information for each VM, but have the authentication portion be performed in the backend. In the future, if we do plan to support data disks / Ultra disks, we can expand this API further to give information about those features too. 

All VM's will support availability zones / sets, as well as managed disks, so this endpoint won't be needed for those attributes. Right now, I'm not sure if we want to either add a new path to list SKU's and their capabilities, or just enhance the VM sizes path to include support for Accelerated networking, lmk what you think and what would be easier for the UI to handle.

**Comment by catherineluse** [comment]
Well, regarding the availability zones - in the Azure docs it said that some VM sizes were not available in some zones, so it had suggested programmatically querying them:

> Some services or VM features are only available in certain regions, such as specific VM sizes or storage types." 

https://learn.microsoft.com/en-us/azure/virtual-machines/regions#feature-availability

I'm not sure why the link in that docs paragraph about programmatically getting the list is going to only the CLI docs without mentioning the API docs, but in any case, it sounds like the VM size availability is rather inconsistent and subject to change.

I would say the existing VM size endpoint in the Norman API should be enhanced to include support for extra filtering.

**Comment by HarrisonWAffel** [comment]
Ah, I must have forgotten, thanks for pointing that documentation out! I've created a ticket in rancher/rancher for this, I'll start working on it now. 

I also wanted to ask if RKE1 is currently using this API, and if not how we should support that UI. I think the VM sizes for RKE1 are stored statically, should we just store feature support statically as well or transition to this API?

**Comment by HarrisonWAffel** [comment]
@catherineluse the PR for enhancing the `aksVMSizes` endpoint has been approved, but I'm not sure when it should be merged as I'm not sure if the change in payload format will break the AKS / Azure UI (my guess is that it will). Do you need me to merge this change to unblock your work, or should we merge both the frontend and backend portions at the same time? 


Additionally, rancher-machine has been bumped in Rancher, so the schema for Azure node templates should now include fields such as `AcceleratedNetworking`.

**Comment by gaktive** [comment]
Backend complete, so they're awaiting UI before handing to QA: https://github.com/rancher/rancher/issues/39672 though they will await for all related Azure AD tickets to be completed before testing begins.

**Comment by gennitdev** [comment]
@HarrisonWAffel yes, please merge it.

**Comment by catherineluse** [comment]
@HarrisonWAffel Yes, please merge it - it will help.

**Comment by HarrisonWAffel** [comment]
@catherineluse I've merged the PR to update the API

**Comment by catherineluse** [comment]
I'm going to split this issues into two parts, one for RKE2/K3s and one for RKE1. This is just so that QA can go ahead and test the former changes separately.

I'll change the title of this issue to be just about RKE2/K3s, and open a new issue for RKE1.

# QA Template
## What was fixed, or what changes have occurred

 - Added  accelerated networking (AN) to Azure node options in the Advanced section for RKE2/K3s cluster provisioning

## Areas or cases that should be tested
- When AN is enabled, the list of VM sizes should show that the VM sizes that don't support it are disabled.
- When AN is enabled, the VM size dropdown should have a tooltip explaining why some options are disabled.
- If you have AN disabled, then select a VM that does not support AN (such as `BasicA0`), then try enable AN, you should see an error that says the VM size is incompatible with AN. A warning also shows next to the VM size dropdown. The warning goes away if any other non-disabled VM is selected or if AN is turned off.

**Comment by thaneunsoo** [comment]
### Test Environment: ###
Rancher version: v2.7-head 8062f1e
Rancher cluster type: HA
Docker version: 20.10

Downstream cluster type: rke2, k3s

----------

### Testing: ###
The configuration is available under `Advanced` section in `Machine Pools` The toggle button is working as expected
![image.png](https://zube.io/files/rancher/ad7e4adb132794200925d8c21bfee44a-image.png)
